### ISSUE TITLE
New 'dynamic' does not extend 'A' warning

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#24713]
When upgrading the SDK to 1.13.0-dev.7.1, a new warning started appearing:

```dart
class A { }
class M<T extends A> { }
class Q { bool test(M arg) => arg is M<dynamic>; }
void main() { }
```

```
Analyzing [test.dart]...
[warning] 'dynamic' does not extend 'A' (test.dart, line 4, col 40)
1 error found.
```

Is this intended?

I'm happy to replace `dynamic` with `A` which is probably better anyway (and would, in unchecked mode, reject things like M<int>, which the current code would not). Just curious if this is intended.

**Comment by Hixie** [comment]
Actually, this is a problem. The actual code I have is more like this:

```dart
class A { }
class M<T extends A> { }

class X<T extends A, U extends M<T>> { }

class Z<V extends M> {
  bool test(X x) => x is X<dynamic, V>;
}

void main() { }
```

There's no way I can see to write that expression without triggering some warning.

**Comment by leafpetersen** [comment]
I don't think this is a strong mode issue - the warning is there without --strong, and my read of the spec is that it shouldn't be a warning in regular Dart.

@ianh, what's wrong with using `x is X<A, V>` there?

**Comment by Hixie** [comment]
That was my first thought, but the analyzer then says V doesn't extend M<A>.

**Comment by leafpetersen** [comment]
Ok, the latter warning *is* strong mode.   Can you write it like this:
```dart
class Z<V extends M<A>> {
  bool test(X x) => x is X<A, V>;
}
```
I think that's a more accurate characterization of the bound, assuming that you actually intend to never instantiate `M` with dynamic (and hence never intended to instantiate `Z` with `M<dynamic>`).

**Comment by Hixie** [comment]
We do sometimes instantiate `Z` with `dynamic`. (When we do, we don't use `test()` or anything else depending on `V`.)

**Comment by Hixie** [comment]
Doing what you suggest does solve the problem for that example, but for some reason doesn't work on the original code. Let me try to minimise it again to see what I missed.

**Comment by Hixie** [comment]
Ok here's a new version of the test (I renamed the classes so that their names here are derived from the original names. W = Widget, E = Element, S = State, G = GlobalKey.)

```dart
abstract class W { }
abstract class E<T extends W> { }
abstract class X extends W { }
abstract class S<T extends X> { }
abstract class F<T extends X, U extends S<T>> extends E<T> {
  U get value;
}

abstract class G<T extends S<X>> {
  T test(E arg) {
    if (arg is F<X, T>) {
      F<X, T> farg = arg;
      return farg.value;
    }
    return null;
  }
}

void main() { }
```

This says:
```
Analyzing [test.dart]...
[hint] Runtime check on non-ground type F<X, T> may throw StrongModeError (test.dart, line 11, col 9)
[error] Type check failed: farg.value (X) is not of type T (test.dart, line 13, col 14)
[error] Type check failed: farg.value (X) is not of type T (test.dart, line 13, col 14)
[warning] The return type 'X' is not a 'T', as defined by the method 'test' (test.dart, line 13, col 14)
4 errors found.
```

However, I just noticed that if I change the name of the type argument in class G from T to U, the errors go away. ?????

**Comment by leafpetersen** [comment]
Ok, that is definitely an analyzer bug, probably in substitution.  Here's a small reproduction case:
```dart
abstract class F<T, U>  {
  U get value;
}

abstract class G<T> {
  T test(F<int, T> arg) => arg.value;
}

abstract class H<S> {
  S test(F<int, S> arg) => arg.value;
}

void main() { }
```
In `G` the regular analyzer gives an error, and in `H` it doesn't.  @jmesserly has been looking at the substitution code a quite a bit lately with an eye to refactoring it to handle generic types - any thoughts about this?

**Comment by jmesserly** [comment]
@leafpetersen -- wow, that's a great reproduction! So somewhere it's getting confused by the name "T", since `G<T>` is (incorrectly?) causing an error. Off hand I don't see how that could happen -- as those T's should be totally unrelated TypeParameterElements/Types. I'll have a look at stepping through it. Thanks for kicking my way, definitely have been stepping through that code a lot lately.

**Comment by jmesserly** [comment]
One random guess -- there's two ways member lookups happen, sometimes they use the Type and other times they use InheritanceManager. The latter one creeps me out, because you'll look for a member starting from a ClassElement. That's creepy because it doesn't track precisely the type substitutions, which seems a bit error prone. There's code [like this](https://github.com/dart-lang/sdk/blob/ba6078197ba8093706873569d123cd94aa897308/pkg/analyzer/lib/src/generated/resolver.dart#L5773) that's tries to do substitution after the fact, to fix things up. (That's not the normal way member types are looked up, BTW ... in fact, it's hardly ever used. But just goes to show similar but slightly different paths where substitutions can happen.)

**Comment by jmesserly** [comment]
It looks like the bug is in StaticTypeAnalyzer._getTypeOfProperty, specifically the code path:
```dart
      // if the return type is a TypeParameter, we try to use the context [that
      // the function is being called on] to get a more accurate returnType type
```
the problem is: we already have a substituted return type at this point. Trying to substitute again is just wrong. That code looks like a hold over from an earlier design. (The fact that we're comparing type parameters by name is a giveaway that something is really wrong here).

**Comment by leafpetersen** [comment]
Yeah, that's an odd looking bit of code.  I wonder if that function was originally designed just to be used to find propagated (inferred) types, and drifted?  Otherwise I don't know what the comment is supposed to mean.  Thanks for tracking that down - is the fix clear, or should we try to grab someone from the analyzer team?

**Comment by jmesserly** [comment]
https://codereview.chromium.org/1409143009/

**Comment by Hixie** [comment]
Do you know when I'll be able to roll flutter to a version with this? (I'm not familiar with the Dart SDK release cycle.)

**Comment by leafpetersen** [comment]
Thanks John!  @sethladd, do you know what branch flutter rolls from? 

cc @dgrove

**Comment by Hixie** [comment]
```
SDK_URL_BASE = ('http://gsdview.appspot.com/dart-archive/channels/dev/raw/'
		'1.13.0-dev.6.0/sdk/')
```

**Comment by leafpetersen** [comment]
I've submitted a merge request for John's fix.  Once that lands (should happen within a day or two), flutter will get the fix once it rolls to the next dev release of the sdk.

**Comment by leafpetersen** [comment]
The fix for the substitution bug has landed in 1.13.0-dev.7.4 .  With this fix, the pattern of using the upper bound as the type argument for is checks should work.  Strong mode will complain if you use dynamic as an argument when the upper bound is tighter than Object/dynamic, regular analyzer doesn't.

**Comment by Hixie** [comment]
excellent, thanks!
